src/tactic/linear_combination.lean

```diff
@@ -0,0 +1,388 @@
+/-
+Copyright (c) 2022 Abby J. Goldberg. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Abby J. Goldberg
+-/
+
+import tactic.ring
+
+/-!
+
+# linear_combination Tactic
+
+In this file, the `linear_combination` tactic is created.  This tactic attempts
+to prove the target by creating and applying a linear combination of a list of
+equalities.  This file also includes a definition for
+`linear_combination_config`.  A `linear_combination_config` object can be
+passed into the tactic, allowing the user to specify a normalization tactic.
+
+## Implementation Notes
+
+This tactic works by creating a weighted sum of the given equations with the
+given coefficients.  Then, it subtracts the right side of the weighted sum
+from the left side so that the right side equals 0, and it does the same with
+the target.  Afterwards, it sets the goal to be the equality between the
+lefthand side of the new goal and the lefthand side of the new weighted sum.
+Lastly, it uses a normalization tactic to see if the weighted sum is equal
+to the target.
+
+## References
+
+* <https://leanprover.zulipchat.com/#narrow/stream/239415-metaprogramming-.2F.20tactics/topic/Linear.20algebra.20tactic/near/213928196>
+
+-/
+
+namespace linear_combo
+
+/-! ### Lemmas -/
+
+lemma left_mul_both_sides {α} [h : has_mul α] {x y : α} (z : α) (h1 : x = y) :
+  z * x = z * y :=
+congr_arg (has_mul.mul z) h1
+
+lemma sum_two_equations {α} [h : has_add α] {x1 y1 x2 y2 : α} (h1 : x1 = y1)
+  (h2: x2 = y2) : x1 + x2 = y1 + y2 :=
+congr (congr_arg has_add.add h1) h2
+
+lemma left_minus_right {α} [h : add_group α] {x y : α} (h1 : x = y) :
+  x - y = 0 :=
+sub_eq_zero.mpr h1
+
+lemma all_on_left_equiv {α} [h : add_group α] (x y : α) :
+  (x = y) = (x - y = 0) :=
+propext (⟨left_minus_right, sub_eq_zero.mp⟩)
+
+lemma replace_eq_expr {α} [h : has_zero α] {x y : α} (h1 : x = 0) (h2 : y = x) :
+  y = 0 :=
+by rwa h2
+
+
+/-! ### Configuration -/
+
+
+/--
+A configuration object for `linear_combination`.
+
+`normalize` describes whether or not the normalization step should be used.
+
+`normalization_tactic` describes the tactic used for normalization when
+checking if the weighted sum is equivalent to the goal (when `normalize` is tt).
+-/
+meta structure linear_combination_config : Type :=
+(normalize : bool := tt)
+(normalization_tactic : tactic unit := `[ring1])
+
+
+/-! ### Part 1: Multiplying Equations by Constants and Adding Them Together -/
+
+
+/--
+Given that lhs = rhs, this tactic returns an expr stating that
+  coeff * lhs = coeff * rhs.
+
+* Input:
+  * `heq` : an expr, which should be an equality with some type α on each side,
```

Comment:
These are all very helpful comments, thank you.  I was not thinking about [docs#heq](https://leanprover-community.github.io/mathlib_docs/init/core.html#heq) when I named those variables.  I can see why that would be confusing, so I updated all of the variables previously named `heq`.  

I also updated the docstrings to more properly describe the equalities, `has_mul`, `has_add`, and `add_group`.  Additionally, I tried to change the wording in several of the docstrings to reflect that they are proving something rather than stating something.